### PR TITLE
Helm template updates for multicluster with gateways

### PR DESCRIPTION
**Pull request by rshriram**
Based on the new filters that went in and some of the recent changes
(sni-dnat mode), I have updated the helm templates for the gateways. 
The istio-mcgateway is no longer needed. We can use the existing 
ingress/egress gateways.

Given that the docs for the multicluster with gateways will require the
user to install some DNS, I have included an addon for the coreDNS plugin
(istio-ecosystem/istio-coredns-plugin) so that users will be able to do
dns resolution for service entries of the form name.namespace.global.
Note that the DNS is optional and users can bring their own DNS.

see associated docs: https://github.com/istio/istio.io/pull/2847

Signed-off-by: Shriram Rajagopalan <shriramr@vmware.com>

### CODE REVIEW

**Comment by rshriram** [comment]
https://github.com/istio/istio.io/pull/2847 docs

File: install/kubernetes/helm/subcharts/gateways/templates/preconfigured.yaml [hunk 1]
```diff
@@ -173,13 +226,41 @@ metadata:
...
    filters:
    - listenerMatch:
        portNumber: 15443
        listenerType: GATEWAY
      filterName: envoy.filters.network.sni_cluster
      filterType: NETWORK
      filterConfig: {}
+   - listenerMatch:
+       portNumber: 15443
+       listenerType: GATEWAY
```
Comment (linsun) [inline install/kubernetes/helm/subcharts/gateways/templates/preconfigured.yaml]: is this a dup of line 231-233?

Comment (linsun) [reply install/kubernetes/helm/subcharts/gateways/templates/preconfigured.yaml]: or a different envoy filter but listen on the same port?

File: install/kubernetes/helm/subcharts/gateways/templates/preconfigured.yaml [hunk 2]
```diff
@@ -128,32 +120,93 @@ spec:
...
+  - "*.global"
+  tls:
   - match:
-    - port: 80
+    - port: 15443
+      sniHosts:
+      - "*.global"
     route:
-    - destination:
+    - destination:
```
Comment (linsun) [inline install/kubernetes/helm/subcharts/gateways/templates/preconfigured.yaml]: remove spaces?

File: install/kubernetes/helm/subcharts/istiocoredns/coredns.yaml [hunk 3]
```diff
@@ -0,0 +1,125 @@
+---
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: istio-coredns-plugin
+  namespace: istio-system
```
Comment (linsun) [inline install/kubernetes/helm/subcharts/istiocoredns/coredns.yaml]: Not hardcode the namespace?

File: install/kubernetes/helm/subcharts/istiocoredns/coredns.yaml [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+---
+# Source: istio/charts/pilot/templates/clusterrole.yaml
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: ClusterRole
+metadata:
+  name: istio-coredns-plugin
+  namespace: istio-system
```
Comment (linsun) [inline install/kubernetes/helm/subcharts/istiocoredns/coredns.yaml]: Not hardcode the namespace?

File: install/kubernetes/helm/subcharts/istiocoredns/coredns.yaml [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+- apiGroups: ["networking.istio.io"]
+  resources: ["*"]
+  verbs: ["get", "watch", "list"]
+---
+# Source: istio/charts/pilot/templates/clusterrolebinding.yaml
+apiVersion: rbac.authorization.k8s.io/v1beta1
+kind: ClusterRoleBinding
+metadata:
+  name: istio-coredns-plugin
+  namespace: istio-system
```
Comment (linsun) [inline install/kubernetes/helm/subcharts/istiocoredns/coredns.yaml]: same namespace issue here  and a few line below

File: install/kubernetes/helm/subcharts/istiocoredns/coredns.yaml [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+kind: ConfigMap
+metadata:
+  name: coredns
+  namespace: istio-system
+data:
+  Corefile: |
+    .:53 {
+        errors
+        health
+        proxy __BEGIN__ .FederationDomain __END__ 127.0.0.1:8053 {
```
Comment (linsun) [inline install/kubernetes/helm/subcharts/istiocoredns/coredns.yaml]: what is the port 8053 for?  would be good to have a comment :)

File: install/kubernetes/helm/subcharts/istiocoredns/coredns.yaml [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+            port: 8080
+            scheme: HTTP
+          initialDelaySeconds: 60
+          timeoutSeconds: 5
+          successThreshold: 1
+          failureThreshold: 5
+      - name: istio-coredns-plugin
+        command:
+        - /usr/local/bin/plugin
+        image: rshriram/istio-coredns-plugin:istio-1.1
```
Comment (linsun) [inline install/kubernetes/helm/subcharts/istiocoredns/coredns.yaml]: could we not use personal image?   should we move coredns plugin to istio repo?

File: install/kubernetes/helm/subcharts/istiocoredns/templates/deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,91 @@
+apiVersion: extensions/v1beta1
+kind: Deployment
```
Comment (linsun) [inline install/kubernetes/helm/subcharts/istiocoredns/templates/deployment.yaml]: you seem to already have the deployment section in coredns.yaml earlier.

Comment (linsun) [reply install/kubernetes/helm/subcharts/istiocoredns/templates/deployment.yaml]: perhaps coredns.yaml should not be part of the PR...

File: install/kubernetes/helm/subcharts/istiocoredns/coredns.yaml [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+  name: istio-coredns-plugin
+  namespace: istio-system
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: istio-coredns-plugin
+subjects:
+  - kind: ServiceAccount
+    name: istio-coredns-plugin
+    namespace: istio-system
```
Comment (morvencao) [inline install/kubernetes/helm/subcharts/istiocoredns/coredns.yaml]: hard-coded namespace, maybe it should be replaced with `{{ .Release.Namespace }}`

File: install/kubernetes/helm/subcharts/istiocoredns/coredns.yaml [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: coredns
+  namespace: istio-system
```
Comment (morvencao) [inline install/kubernetes/helm/subcharts/istiocoredns/coredns.yaml]: hard-coded namespace

File: install/kubernetes/helm/subcharts/gateways/templates/preconfigured.yaml (hunk 1 above)
Comment (rshriram) [reply install/kubernetes/helm/subcharts/gateways/templates/preconfigured.yaml]: different filter

File: install/kubernetes/helm/subcharts/istiocoredns/coredns.yaml (hunk 7 above)
Comment (rshriram) [reply install/kubernetes/helm/subcharts/istiocoredns/coredns.yaml]: We can, if others are okay with it.

**Comment by linsun** [comment]
LGTM.  The other concern is the docker image for core dns plugin in shriram's docker repo.   If that could be addressed in a followup PR soon in 1.1, it'd be okay to get this one in since this is disabled by default.

**Comment by linsun** [comment]
/LGTM

@rshriram let's push your core dns plugin image to istio repo somewhere before 1.1 is shipped.

cc @hklai @costinm

**Comment by rshriram** [comment]
@costinm docs are here:  https://github.com/istio/istio.io/pull/2847
The global part requires more work. Thats the reason I restricted it completely to a separate profile with separate values.yaml. This affects only users that are trying to install the multicluster with gateway mode. To make the docs easy to follow and use and serve as a reference implementation, I had to pick something. Otherwise, its a totally halfbaked implementation like the existing multicluster without gateways - where users do not know how what is necessary to get a basic working system in place before they can go about customizing this.

**Pull request merged**